pkg/controller/podgc/gc_controller.go

```diff
@@ -122,40 +131,32 @@ func (gcc *PodGCController) gcTerminated(pods []*v1.Pod) {
 	}
 	if deleteCount > 0 {
 		klog.Infof("garbage collecting %v pods", deleteCount)
+		var wait sync.WaitGroup
+		for i := 0; i < deleteCount; i++ {
+			wait.Add(1)
+			go func(namespace string, name string) {
+				defer wait.Done()
+				if err := gcc.deletePod(namespace, name); err != nil {
+					// ignore not founds
+					utilruntime.HandleError(err)
+				}
+			}(terminatedPods[i].Namespace, terminatedPods[i].Name)
+		}
+		wait.Wait()
 	}
-
-	var wait sync.WaitGroup
-	for i := 0; i < deleteCount; i++ {
-		wait.Add(1)
-		go func(namespace string, name string) {
-			defer wait.Done()
-			if err := gcc.deletePod(namespace, name); err != nil {
-				// ignore not founds
-				defer utilruntime.HandleError(err)
-			}
-		}(terminatedPods[i].Namespace, terminatedPods[i].Name)
-	}
-	wait.Wait()
 }
 
 // gcOrphaned deletes pods that are bound to nodes that don't exist.
 func (gcc *PodGCController) gcOrphaned(pods []*v1.Pod) {
 	klog.V(4).Infof("GC'ing orphaned")
-	// We want to get list of Nodes from the etcd, to make sure that it's as fresh as possible.
```

Comment:
QQ, now that we have the watch on nodes, cannot we change the approach and instead of calling this method periodically call this method every time we get a "node deleted" event to see if there aren't any pods that became orphaned? I might be wrong, but that should guarantee eventual consistency and would be simpler than what you have in the FileterDeletedNodeNames, WDYT?

Reply:
Implementation could probably be changed to use only watches. I am not sure if that would be substantially simpler. Regarding your exact question:
1. We need to handle restarting controller, when there is no access to old deleted events
2. Do we care about a time between pod becomes orphaned and its removal? Both in case of node being deleted and pod assigned to a not-ever-existing node.
3. Other methods in this controller run periodically. Is there any benefit of running this method only on node deletion (assuming using only data from watches)?